pkg/controller/statefulset/stateful_set.go

```diff
@@ -142,15 +148,40 @@ func (ssc *StatefulSetController) Run(workers int, stopCh <-chan struct{}) {
 func (ssc *StatefulSetController) addPod(obj interface{}) {
 	pod := obj.(*v1.Pod)
 	glog.V(4).Infof("Pod %s created, labels: %+v", pod.Name, pod.Labels)
-	set := ssc.getStatefulSetForPod(pod)
-	if set == nil {
+
+	if pod.DeletionTimestamp != nil {
+		// on a restart of the controller manager, it's possible a new pod shows up in a state that
+		// is already pending deletion. Prevent the pod from being a creation observation.
+		ssc.deletePod(pod)
 		return
 	}
-	ssc.enqueueStatefulSet(set)
+
+	// If it has a ControllerRef, that's all that matters.
+	if controllerRef := controller.GetControllerOf(pod); controllerRef != nil {
+		if controllerRef.Kind != ControllerKind.Kind {
+			// It's controlled by a different type of controller.
+			return
+		}
+		set, err := ssc.setLister.StatefulSets(pod.Namespace).Get(controllerRef.Name)
+		if err != nil {
+			// Don't log NotFound. That just means the GC will soon delete the Pod.
+			if !errors.IsNotFound(err) {
```

Comment:
The cache returns only 404s:
https://github.com/kubernetes/kubernetes/blob/77733c2afdcc4e85595e990d69d9b2765b1cf4e1/pkg/client/listers/apps/v1beta1/statefulset.go#L92
https://github.com/kubernetes/kubernetes/blob/77733c2afdcc4e85595e990d69d9b2765b1cf4e1/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L392